Playwright: migrate coblocks tests.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes to migrate the CoBlocks tests to run under Playwright.

Notable changes:

- implementation of a BaseBlock class from which all Block classes inherit.
- concrete <insertname>Block classes to represent the individual blocks and unique actions that can be performed on them. For instance, `PricingTableBlock` is able to modify gutter width and this is implemented as a member method.
- modifications to the `GutenbergEditorPage` to support insertion of blocks.
- changes to DataHelper to permit user control on where the `parallel` tag should be appended to (parent or child suite).

The following tests are migrated:
- [x] click to tweet
- [x] dynamic HR
- [x] hero
- [ ] logos
- [x] pricing table
- [x] gutter

#### Testing instructions
TeamCity
- ensure tests pass from the `wp-calypso__coblocks-spec`.

worldomonation committed:
Start new branch to migrate coblock tests.

Include test file

Parametrize additional tests.

worldomonation committed:
- address eslint.
- migrate the WPCOM gutter test.
- add new method `setGutter` as member method of PricingTableBlock.

- parallelize each coblock test.
- new function DataHelper.createSubsuiteTitle to generate formatted subsuite test name.

Linter fixes.
- fill JSDoc for new methods.
- remove use of `any` within inline functions.

Large rewrite of the coblock tests.
- split the twitter tests from rest due to an existing issue that auto-dismisses the publish panel.
- moved gutter tests to its own file.
- implemented files for each block.
- implemented a static method to check on the published page whether the block is present.

worldomonation commented:
The logos test should not be migrated at this time as it relies on generating a new test image - which is implemented in #53828, but that PR needs a rework.

worldomonation commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -113,15 +113,44 @@ export function randomPhrase(): string {
+ * @param {boolean} param0.parallel Whether to include the @parallel tag. Defaults to false.
+ * @returns {string} Test subsuite name.
+ */
+export function createSubsuiteTitle(
```
Changes in this file are meant to provide better control to test writers on exactly where the `@parallel` tag should be placed.

Previously the `@parallel` tag was always placed at the parent suite level.

WunderBart replied:
We've moved away from parallel sub-suites - see https://github.com/Automattic/wp-calypso/pull/52914 and https://github.com/Automattic/wp-calypso/pull/52916. This should not be needed. If a parallel sub-suite is needed, it should be created in a separate file.

worldomonation replied:
I had thought we still support subsuite with `@parallel`; seeing that's such is not the case, I will look into splitting these cases.

worldomonation replied:
Removed all subsuite with `@parallel`.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -166,6 +176,61 @@ export class GutenbergEditorPage extends BaseContainer {
+	 *
+	 * @returns {Promise<void>} No return value.
+	 */
+	async openSidebar(): Promise< void > {
```
The name `sidebar` may not be the best here, but I could not glean any useful information from the CSS selector. Any thoughts?

WunderBart replied:
It seems to be called just "Settings", so let's call it that: `openSettings`.

worldomonation replied:
Sounds good.

worldomonation commented on test/e2e/specs/specs-playwright/wp-calypso__coblocks-spec.js:
```diff
@@ -0,0 +1,168 @@
+} from '@automattic/calypso-e2e';
+
+describe( DataHelper.createSuiteTitle( 'Gutenberg: CoBlocks', { parallel: false } ), function () {
+	describe( DataHelper.createSubsuiteTitle( 'Click to Tweet', { parallel: true } ), function () {
```
Note the use of the new method here.

The alternative, to use a simple string for the name is also acceptable, in case reviewer(s) think creating an entire method for this is unnecessary:

`describe( 'Click to Tweet: @parallel') ...`

WunderBart replied:
See my previous comment about why we shouldn't be doing that. Let's split those into separate files.

WunderBart commented on packages/calypso-e2e/src/lib/blocks/pricing-table-block.ts:
```diff
@@ -0,0 +1,44 @@
+
+		const selector = `${ selectors.gutterControl } button[aria-label="${ value }"]`;
+		await frame.click( selector );
+		await frame.$eval( selector, ( element ) => element.getAttribute( 'aria-pressed' ) === 'true' );
```
`$eval` only returns the value of the given page function, so we're not checking anything this way. It will also not wait, so that's another reason to not use it like that. Let's replace it with wait-type call.

worldomonation replied:
I have changed this statement to:

```
return await frame.$eval( selector, ( element ) => element.getAttribute( 'aria-pressed' ) === 'true' );
```

and changed the test step to assert that `is-pressed` attribute is indeed set for the correct gutter value.

WunderBart commented on packages/calypso-e2e/src/lib/base-block.ts:
```diff
@@ -0,0 +1,23 @@
+import { ElementHandle } from 'playwright';
+
+/**
+ * Base class for asynchronously initializing objects.
```
How are we initializing asynchronously here?

worldomonation replied:
Ah this was left over from the copy-paste.

WunderBart commented on packages/calypso-e2e/src/lib/base-block.ts:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Base class for asynchronously initializing objects.
+ */
+export class BaseBlock {
```
What do you think about `GutenbergBlock` instead?

worldomonation replied:
Would this also cover `coblocks` and other types?

WunderBart replied:
I'm thinking yes since all CoBlocks' blocks are essentially Gutenberg blocks. We'd never need a separate base class for Gutenberg and CoBlocks blocks. Does that make sense? Having said that, I can now see that we've already used the `Base*` naming convention for the `BaseContainer` class, so let's stay consistent and keep the prefix as-is.

worldomonation replied:
Agreed, let's leave it as-is.

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -19,12 +19,22 @@ const selectors = {
 
 	// Top bar selectors.
 	publishPanelToggle: '.editor-post-publish-panel__toggle',
+	sidebarToggle: '[aria-label="Settings"]',
```
```suggestion
	settingsToggle: '[aria-label="Settings"]',
```

worldomonation replied:
Done!

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -166,6 +176,61 @@ export class GutenbergEditorPage extends BaseContainer {
+		if ( ! isSidebarOppen ) {
+			await this.frame.click( selectors.sidebarToggle );
+		}
+		await this.frame.waitForSelector( selectors.sidebar );
```
Let's only wait for the `settingsToggle`'s `aria-expanded` attribute to be `"true"`. No need to depend on a wrapper class that can potentially be used somewhere else, compromising this check.

worldomonation replied:
Changed this statement to the following:

```
		const settingsToggle = await this.frame.waitForSelector( selectors.settingsToggle );
		await this.frame.waitForFunction(element => element.getAttribute('aria-pressed') === 'true', settingsToggle)
```

In order to perform the call within `waitForFunction` the argument must be of ElementHandle type.

WunderBart commented on packages/calypso-e2e/src/lib/pages/published/published-post-page.ts:
```diff
@@ -91,4 +94,15 @@ export class PublishedPostPage extends BaseContainer {
 		await this.frame.waitForSelector( selectors.notLikedText, { state: 'visible' } );
 	}
+
+	/**
+	 * Confirms the presenced of the block on the published page.
+	 *
+	 * This method accepts a selector referencing the block to verify whether the block(s) are rendered on page.
+	 *
+	 * @param {string} blockSelector Selector to block(s).
+	 */
+	async confirmBlockPresence( blockSelector: string ): Promise< void > {
+		await this.page.waitForSelector( `${ selectors.postBody } ${ blockSelector }` );
+	}
```
This should be a static method of an individual block, not a general, block-unspecific one. Blocks can differ in a way we'd like them to be verified on the published page - i.e. Image block should also perform a check that the image has been rendered. For example, if I'd like to verify the presence of the `ClickToTweet` block, I'd rather see it called via `await ClickToTweet.validatePublishedContent()` that would already include the `.wp-block-coblocks-click-to-tweet` class, as well as the checks for tweeted text, etc. Does that make sense?

worldomonation replied:
That sounds like a good approach.

worldomonation replied:
I've implemented a static method in each instance of the block to perform checks.

What I'd like to know is if there's a way to generalize this some more. Currently, each block needs to have its own `static validatePublishedContent()` method and this can be problematic if we need to change how the function operates. Do you have any ideas to improve the current implementation?

WunderBart replied:
I think this is OK as-is since each block is different and thus requires a different validation logic. For those checks, I'd focus more on the content that the block produces than on checking if the wrapper element is present. For example, for the `ClickToTweet` block, I think it would bring more value if we confirm that the tweet text is present:

```js
static async validatePublishedContent( page,  tweet ) {
	await page.waitForSelector( `text=${ tweet }` );
}
```

This way we don't even need to even use the block's selector as the text check is sufficient for validating that the block produced expected content.

Similarly, with the `PricingTable` block, it would be great to validate the actual content we published - plan names, prices, features, etc.

What do you think?

worldomonation replied:
I'm in two minds about that, personally.

On one hand, validating the presence of the selectors ought to be enough that the block is present. This covers situations like with the Pricing Table block where if pricing data was not entered the block selector itself would not show up on the published page. Bonus, this keeps the `validatePublishedContent` signature consistent across implementations and manageable.

On the other hand checking for the content to be present would definitely help with ensuring the appropriate values are shown on the block after publishing. With my short time at a8c so far I haven't run into a situation like this, but I'm sure it can happen with a new block or significantly updated block. This would potentially make the method signature very long and complex if the list of things to validate on grows:

```
validatePublishedContent( page, price1, price2, feature1, feature2, plan1name, plan2name, ...)
```

worldomonation replied:
With all that I've typed up, I ended up modifying `validatePublishedContent` to check for values. Please take a look and let me know if it should be changed.

dpasque replied:
Because this function seems to be intended to be used for polymorphism (in the looping test), do you think it's worth it to try to collapse the signatures down to a single signature and include that as part of an interface that each Block implements? That way, tests in TypeScript that consume these blocks could key off that interface and have more strict type checking. 

It would need to no longer be static, but I think that could end up fine. We could also, for each block, define two POM classes per module, one to represent the block in the editor, and another to represent the block in a published page. The former could take the block `ElementHandle` as its dependency in its constructor, and the latter could take the `page` in its constructor. The latter would be what would implement the interface with this method.

worldomonation replied:
> We could also, for each block, define two POM classes per module, one to represent the block in the editor, and another to represent the block in a published page.

Interesting idea and worth exploring some more. It would provide better separation of concerns and method calls tailored for each use case.

My concern would be around overhead for both developers implementing a new block in this module as well as test writers. I would imagine test writers needing to context switch between the editor and published representations of the block:

```
PricingTableBlock.something

...

PricingTableBlockPublished.validate()
```

whereas currently the test writer only needs to know that `PricingTableBlock` can handle both use cases.

worldomonation replied:
>  do you think it's worth it to try to collapse the signatures down to a single signature and include that as part of an interface that each Block implements?

Would this work for blocks that require no validation eg. Dynamic HR block? The validation method there only checks whether the class exists on page.

worldomonation replied:
Personally, I think we should hold off on making this change for now and perhaps look at revising this when additional block tests are migrated. It should not be too difficult to make the changes at that time.

dpasque replied:
Sounds good to me! We can definitely revisit this.

WunderBart commented on test/e2e/specs/specs-playwright/wp-calypso__coblocks-spec.js:
```diff
@@ -0,0 +1,168 @@
+			gutenbergEditorPage = await GutenbergEditorPage.Expect( this.page );
+		} );
+
+		it( 'Enter post title', async function () {
+			await gutenbergEditorPage.enterTitle( blockName );
+		} );
```
Is this step necessary?

worldomonation replied:
Since we are publishing the post, I think it would be better to have posts with titles.

worldomonation committed:
Rebase and use Jest.

worldomonation committed:
Update the expected CSS for Dyanmic HR block.
Update JSDoc for return value on block classes.

worldomonation committed:
Placate the linter

WunderBart commented on packages/calypso-e2e/src/lib/blocks/pricing-table-block.ts:
```diff
@@ -0,0 +1,56 @@
+	 * @returns {Promise<void>} No return value.
+	 */
+	async setGutter( value: 'None' | 'Small' | 'Medium' | 'Large' | 'Huge' ): Promise< boolean > {
+		const frame = ( await this.block.ownerFrame() ) as Frame;
```
This looks odd, can we initialize the Block class with the parent page (as we do with the `BaseComponent`) and use that instead?

worldomonation replied:
Right - my intention with `BaseBlock` class is to focus solely on the block itself where possible, but the gutter portion sneaks out from that and requires interacting with the larger page.

With that said though, keep in mind that we're interacting with the iframed Gutenberg editor, so `page` won't work here. Calling `page.click` while on the editor will fail to register actions. 
So I'd say retaining this approach is the best approach, and it is similar to how `GutenbergEditorPage` operates as well.

WunderBart replied:
> Right - my intention with BaseBlock class is to focus solely on the block itself where possible, but the gutter portion sneaks out from that and requires interacting with the larger page

Gotcha. I can see how that makes the responsibility more focused, but at the same time, I think that the block settings panel is as much a block as the block element itself. Does that make sense?

> With that said though, keep in mind that we're interacting with the iframed Gutenberg editor, so page won't work here. Calling page.click while on the editor will fail to register actions.

Thanks for the explanation, I've missed that one. In some future PR, I think it would make sense to make a base method for switching to the block settings. 

> So I'd say retaining this approach is the best approach, and it is similar to how GutenbergEditorPage operates as well.

Sounds good! 👍

WunderBart commented on packages/calypso-e2e/src/lib/blocks/pricing-table-block.ts:
```diff
@@ -0,0 +1,56 @@
+		const frame = ( await this.block.ownerFrame() ) as Frame;
+
+		const selector = `${ selectors.gutterControl } button[aria-label="${ value }"]`;
+		await frame.click( selector );
+		return await frame.$eval(
+			selector,
+			( element ) => element.getAttribute( 'aria-pressed' ) === 'true'
+		);
```
The `$eval` call is not a `wait`-type call so it's possible to get a flaky result here. Do you think it would be better to  make this helper self-validate instead?

```suggestion
		const targetButton = await frame.waitForSelector( selector );
		await targetButton.click();
		await page.waitForFunction( ( element: any ) => element.ariaPressed === 'true', elementHandle ),
```

worldomonation replied:
Updated this call to be the following:

```
		await frame.click( selector );
		const elementHandle = await frame.waitForSelector( selector );
		await frame.waitForFunction(
			( element: any ) => element.ariaPressed === 'true',
			elementHandle
		);
```

WunderBart commented on test/e2e/specs/specs-playwright/wp-calypso__coblocks-click-to-tweet-spec.js:
```diff
@@ -0,0 +1,58 @@
+	} );
+
+	it( `Click to Tweet block is visible in published post`, async function () {
+		await ClicktoTweetBlock.validatePublishedContent( page );
```
I'm a bit confused here - I thought we'd need a new page instance since we're publishing - my bad. If we don't really need that, I think we should pass the current `page` along when initializing the block and make the `validatePublishedContent` an instance method. What do you think? Sorry for going back and forth on this one - still trying to get my mind back on track after the break 😄

worldomonation replied:
Ah yeah - the underlying `Page` object is retained throughout the test, until the `page` is closed or terminated in some way.

worldomonation replied:
Although, I think keeping this as a `static` method would make more sense; let me know if my line of reasoning is sound here.

A static class method it isn't bound to a specific instance of the class:
- when validating that a block exists on a published page, we are no longer working with the block itself on the editor, which is what an instance of a Block class represents.
- there may be multiple blocks of the same type and we may want to validate all of them, by checking for different text value.

WunderBart replied:
Good points! Let's keep it static 👍

WunderBart commented on test/e2e/specs/specs-playwright/wp-calypso__coblocks-click-to-tweet-spec.js:
```diff
@@ -0,0 +1,58 @@
+
+	it( 'Publish and visit post', async function () {
+		await gutenbergEditorPage.publish( { visit: true } );
+		await PublishedPostPage.Expect( page );
```
Should this be a part of the `publish` method (when the `visit` is set to true)?

worldomonation replied:
Actually I think that call isn't necessary at all and will be removed in next push.

WunderBart commented on packages/calypso-e2e/src/lib/blocks/pricing-table-block.ts:
```diff
@@ -0,0 +1,56 @@
+	 * @param {string} price Price to be entered.
+	 * @returns {Promise<void>} No return value.
+	 */
+	async enterPrice( side: 'left' | 'right', price: string | number ): Promise< void > {
+		const index = side === 'left' ? 1 : 2;
+		const priceHandler = await this.block.waitForSelector(
+			`:nth-match(${ selectors.pricing }, ${ index })`
+		);
+		await priceHandler.fill( price.toString() );
+	}
```
nit: We can make it a bit simpler:
```suggestion
	async enterPrice( column: number, price: string | number ): Promise< void > {
		const priceHandler = await this.block.waitForSelector(
			`:nth-match(${ selectors.pricing }, ${ column })`
		);
		await priceHandler.fill( price.toString() );
	}
```

Do you think we should also make it self-validate?

worldomonation replied:
I do not think this method should validate - that is the job of the `validatePublishedContent`.

worldomonation replied:
I've updated the method to the slightly simpler one you suggested. I think the column number works better as well, due to me finding out now that it's possible to alter the number of columns. That might be a good future test.

WunderBart replied:
> I do not think this method should validate - that is the job of the `validatePublishedContent`.

I think there are two sides to this:
1. Making sure that the content has actually been entered into the block when editing, and
2. Making sure that the content is rendered correctly in the front-end.

It's possible for the content to be broken in the editor but OK in the front-end, and vice-versa (see https://github.com/Automattic/wp-calypso/issues/54276). When we make the editor methods self-validate, IMO we're kinda covering the part where we'd need to have separate tests for these methods ( e.g. "enterPrice fills the price correctly"). Having said that, I think that in this case the `fill` method's actionability checks are enough of a self-validation, so it's all good here! 😄

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -143,6 +154,60 @@ export class GutenbergEditorPage extends BaseContainer {
+	 * @returns {Promise<void>} No return value.
+	 */
+	async openSettings(): Promise< void > {
+		const isSidebarOppen = await this.frame.$eval( selectors.settingsToggle, ( element ) =>
```
```suggestion
		const isSidebarOpen = await this.frame.$eval( selectors.settingsToggle, ( element ) =>
```

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -143,6 +154,60 @@ export class GutenbergEditorPage extends BaseContainer {
+		const isSidebarOppen = await this.frame.$eval( selectors.settingsToggle, ( element ) =>
+			element.classList.contains( 'is-pressed' )
+		);
+		if ( ! isSidebarOppen ) {
```
```suggestion
		if ( ! isSidebarOpen ) {
```

WunderBart commented on test/e2e/specs/specs-playwright/wp-calypso__coblocks-gutter-spec.js:
```diff
@@ -0,0 +1,51 @@
+
+	it.each`
+		gutterValue
+		${ 'None' }
+		${ 'Small' }
+		${ 'Medium' }
+		${ 'Large' }
+		${ 'Huge' }
```
Can we have these assigned to the static prop of the `PricingTableBlock` (e.g. `PricingTableBlock.gutterValues`) and [map them](https://jestjs.io/docs/api#1-testeachtablename-fn-timeout) in this `each` block instead of re-defining them?

worldomonation replied:
I'd like clarification here when you refer to `prop` - is the suggestion to make a new property in a class, like this?

```
class PricingTableBlock extends ... {
    name: PricingTableBlock
    gutterValues: [
         ....
    ]
```

worldomonation replied:
I spent some time trying to do this, but was unable to come to a solution I found satisfactory.

These are the approaches I tried:
1. move the gutter value definition to a static property in the Block file.
2. move the gutter value definition to the type definition file then import that in spec file.

For approach 1, I was unable to have the gutter values defined as a static property in the Block class to be used as the type definition in the argument of `setGutter` method. So there needed to be two locations where this was defined, defeating the purpose.

For approach 2, import of type definition appears to become `undefined` in the spec file. I wonder if it has to do with the JS-TS duality we have in the calypso e2e ecosystem. 

Pinging TS expert @dpasque as he might have some ideas here?

WunderBart replied:
Would something like this work in this case?

```ts
const gutterValues = [ 'Small', 'Large' ] as const;
type GutterValues = typeof gutterValues[ number ];

class TheBlock {
    static gutterValues = gutterValues

    setGutter( value: GutterValues ) {
        // set the gutter
    }
}
```

WunderBart replied:
Ah, I can see now you kinda did the same thing, although moved the type definition out of the block module. Is it OK if we just leave the definitions contained as in the example above?

dpasque replied:
My 0.02 on these kind of approaches:

- I very much agree with Bart that I think it makes the most sense for these typings to live in the same modules as the block classes they are for (unless they exist more broadly). You can define and export them at the top of the module outside the class.
- If you only need type safety for the types of strings you are passing as an argument, you can just a union string type. I often like to export this type because it helps TypeScript files that consume that type get the right intellisense and be able to use that type for local variables.
```typescript
export type GutterValues = 'Small' | 'Large'
```
- If you want people to be able to interact with an array of the full set of these values, they way you are currently doing it is a great way. In that case though, I would also make sure to export the constant array as well! That way folks can use it and loop on it. This also gives some value to JavaScript file consumers. Types aren't included in transpiled JS, so that's why you were seeing the `undefined`s when you were importing the type. However, you could just import that array and loop over it directly in the test!

worldomonation replied:
Thanks for input, I had thought the best practice was to group all type definitions in one file but it doesn't seem that's the case.

I've made the following changes:

1. move the type and array definition of the gutter values to the PricingTableBlock.
2. assign a static property `gutterValues` to the PricingTableBlock class using the array defined in 1.
3. use the static property `gutterValues` to build the parametrized test case in the spec.

WunderBart commented on test/e2e/specs/specs-playwright/wp-calypso__coblocks-spec.js:
```diff
@@ -0,0 +1,69 @@
+	} );
+
+	it.each`
+		block                  | name
+		${ PricingTableBlock } | ${ 'Pricing Table' }
+		${ DynamicHRBlock }    | ${ 'Dynamic HR' }
+		${ HeroBlock }         | ${ 'Hero' }
```
Ditto. IMO, the block name should also be a static property of the block class. This way we create a single source of truth and only one place that will need to be updated in case anything changes.

worldomonation replied:
Done!

worldomonation committed:
PR comments.
- use `waitForFunction` when setting gutters.
- fix typo in GutenbergEditorPage.
- remove header from BaseBlock class.
- simplify column selection in PricingTableBlock.
- remove unnecessary page objects from all coblock tests.

scinos commented on packages/calypso-e2e/src/lib/base-block.ts:
```diff
@@ -0,0 +1,23 @@
+/**
```
`packages/calypso-e2e` was migrated to use `import/order` a couple of weeks ago (#54542). You don't need to specify these import dock blocks anymore.

worldomonation replied:
This PR originated before the changeover, but I will make sure to remove these headers in the next revision.

worldomonation committed:
Major changes to address code review comments:
- use static property in the Block classes in test specs.
- static `validatePublishedContent` method to check for content entered into the block instead of asserting on wrapper classes.
- merged Click to Tweet test into the main coblock test suite.

worldomonation committed:
Define gutter values in types.

worldomonation committed:
Add proper mobile support.

worldomonation commented:
Look at that, all tests passed on Playwright in first run (incl. mobile)!

WunderBart commented on test/e2e/specs/specs-playwright/wp-calypso__coblocks-spec.js:
```diff
@@ -0,0 +1,82 @@
+		await gutenbergEditorPage.enterTitle( DataHelper.randomPhrase() );
+	} );
+
+	it( `Insert ${ PricingTableBlock.name } block and enter price to left table`, async function () {
```
Typo?
```suggestion
	it( `Insert ${ PricingTableBlock.blockName } block and enter price to left table`, async function () {
```

worldomonation replied:
Thanks

dpasque commented on packages/calypso-e2e/src/lib/base-block.ts:
```diff
@@ -0,0 +1,20 @@
+import { ElementHandle } from 'playwright';
+
+/**
+ * Base class for the Block type objects.
+ */
```
This is definitely optional as it's just stylistic (and personal preference), but I wanted to throw it out there...

How do you feel about ditching the inheritance model here? With the new approaches to the `BaseContainer`, we're so close to being completely inheritance-free in this new Playwright infrastructure! 

My basic pitch is this: I think if we can avoid inheritance, it will help the code grow in a more readable and newcomer-friendly way. I think it can make these Block classes more explicit about how they work, and make it clearer for new folks on where they should add code (and avoiding split implementations, overrides, etc.).

The base constructor is just declaring what dependencies it needs (the element handle), and I think that's not too verbose to have each Block class just declare that, especially because the dependencies may vary across block (e.g., maybe a Block needs both a pointer to the Block element handle _and_ the Playwright `page`).

In the future, if we want shared functionality that applies to all blocks, I think we can make a separate set of library functions, and if needed have those library functions consume interfaces that we can then apply to relevant blocks. 

What do you think?

worldomonation replied:
Oh that's a good point, I will remove inheritance in the next commit.

dpasque commented on packages/calypso-e2e/src/lib/base-block.ts:
```diff
@@ -0,0 +1,20 @@
+ */
+export class BaseBlock {
+	[ x: string ]: any;
+	static block: ElementHandle;
```
If we do keep this base class, I think we don't want this to be static, right? This should be unique to each instance of this class?

worldomonation replied:
Ah, really good catch - thanks! This would have had an unintentional side effect when more than 1 block of the same type was to be inserted into the same editor, I assume.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -143,6 +154,60 @@ export class GutenbergEditorPage extends BaseContainer {
+	 *
+	 * @param {string} blockName Name of the block to be inserted.
+	 */
+	async addBlock( blockName: string ): Promise< ElementHandle > {
+		// Click on the editor title. This has the effect of dismissing the block inserter
+		// if open, and restores focus back to the editor root container, allowing insertion
+		// of blocks.
+		await this.frame.click( selectors.editorTitle );
+		await this.openBlockInserter();
+		await this.frame.fill( selectors.blockSearch, blockName );
+		await this.frame.click( `${ selectors.blockInserterResultItem }:has-text("${ blockName }")` );
+		// Confirm the block has been added to the editor body.
+		return await this.frame.waitForSelector( `[aria-label="Block: ${ blockName }"]` );
+	}
```
I just wanted to say I really like this pattern you've locked in where the `GutenbergEditorPage` serves almost as a factory to create the ElementHandles we then feed as dependencies for Block POMs. It reads really well and I think it's going to grow and scale fantastically!

worldomonation replied:
Hey look, unintentional factory pattern 😆

fullofcaffeine replied:
This is a great improvement over the existing `addBlock` method, I like how concise it is!

dpasque commented on test/e2e/specs/specs-playwright/wp-calypso__coblocks-spec.js:
```diff
@@ -0,0 +1,82 @@
+		await gutenbergEditorPage.publish( { visit: true } );
+	} );
+
+	it.each`
+		block                  | content
+		${ PricingTableBlock } | ${ [ pricingTableBlockPrice ] }
+		${ DynamicHRBlock }    | ${ null }
+		${ HeroBlock }         | ${ [ heroBlockHeading ] }
+		${ ClicktoTweetBlock } | ${ [ clicktoTweetBlockTweet ] }
+	`(
+		`Confirm $block.blockName block is visible in published post`,
+		async ( { block, content } ) => {
+			// Passing in the actual object reference permits this call to succeed.
+			// Calling `eval(blockName)` or `global[blockName]` does not work here.
+			await block.validatePublishedContent( page, content );
+		}
+	);
```
This loop turned out so slick! I hope we can continue to do more things like this as we keep migrating!

dpasque reviewed (commented):
Offered a few comments and my 0.02 on some of the questions! Overall looking really nice!

worldomonation committed:
Address review comments.
- modify some comments in code.
- move type definition for gutterValues to the PricingTableBlock itself.
- parametrize the gutter value check using values defined in PricingTableBlock.

worldomonation committed:
Remove inheritance from blocks

WunderBart reviewed (approved):
🚀

dpasque commented on packages/calypso-e2e/src/lib/base-block.ts:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Base class for the Block type objects.
+ */
+export class BaseBlock {
```
Should we remove this file from the commit because it is no longer needed?

worldomonation replied:
I thought I had removed it, my bad.

worldomonation replied:
Done!

dpasque reviewed (approved):
These changes look awesome! My one recommendation would be to remove the BaseBlock class and file now that it is no longer used, but otherwise LGTM and ready to 🚢 !

worldomonation committed:
Remove BaseBlock file.
